### PR TITLE
Progressively retrying uploads to bytestream

### PR DESCRIPTION
**Pull request by werkt**
Attempt to continue uploads through retriable exceptions, using the
ProgressiveBackoff to reset as progress is made. This adds a write
status query after each retriable exception to determine whether
progress has been made. A service that does not support write
resumption, either with persistent write resets or failure status
responses, the behavior is identical to the non-progressive
implementation.

The ProgressiveBackoff was also added to the read side for executeAsync
invocations to ensure that it was used instead of the retrier supplier,
and the injected backoff executeAsync Retrier method was exposed.

### CODE REVIEW

**Comment by buchgr** [comment]
thanks a lot for the PR! I unfortunately won't get to this until next week :-(.

File: src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java [hunk 1]
```diff
@@ -83,7 +89,7 @@
   private final Set<HashCode> uploadedBlobs = new HashSet<>();
 
   @GuardedBy("lock")
-  private final Map<Digest, ListenableFuture<Void>> uploadsInProgress = new HashMap<>();
+  private final Map<HashCode, ListenableFuture<Void>> uploadsInProgress = new HashMap<>();
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java]: Why do you need to change the types here?

File: src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java [hunk 2]
```diff
@@ -260,26 +264,34 @@ boolean uploadsInProgress() {
...
-  private ListenableFuture<Void> startAsyncUpload(Chunker chunker) {
+  private ListenableFuture<Void> startAsyncUpload(HashCode hash, Chunker chunker) {
     try {
       chunker.reset();
     } catch (IOException e) {
       return Futures.immediateFailedFuture(e);
     }
 
-    SettableFuture<Void> currUpload = SettableFuture.create();
+    UUID uploadId = UUID.randomUUID();
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java]: move this into newResourceName?

File: src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java [hunk 3]
```diff
@@ -309,43 +321,157 @@ public ReferenceCounted touch(Object o) {
...
+      // FIXME we should also return immediately without the query if
+      // we were out of retry attempts for the underlying backoff. This
+      // is meant to be an only in-between-retries query request.
+      if (!retrier.isRetriable(e)) {
+        return exceptionFuture;
+      }
+
+      ListenableFuture<Void> suppressedQueryFuture = Futures.catchingAsync(
+          query(committedOffset, progressiveBackoff),
+          Throwable.class,
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java]: Why Throwable? We shouldn't be catching errors.

File: src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java [hunk 4]
```diff
@@ -309,43 +321,157 @@ public ReferenceCounted touch(Object o) {
...
+              QueryWriteStatusRequest.newBuilder()
+                  .setResourceName(resourceName)
+                  .build()),
+          (response) -> response.getCommittedSize(),
+          MoreExecutors.directExecutor());
+      return Futures.transformAsync(
+          committedSizeFuture,
+          (committedSize) -> {
+            if (committedSize > committedOffset.get()) {
+              progressiveBackoff.reset();
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java]: add a comment on why we reset the backoff?

**Comment by buchgr** [comment]
@werkt can you please also special case the "NOT_IMPLEMENTED" status code for QueryWriteStatus to be treated as if written_bytes=0? This would be great for bk as some backends might not implement this method.

File: src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java (hunk 1 above)
Comment (werkt) [reply src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java]: Once I'd implemented the progressive upload, I found I had no need for the digest size, and factored it out. It also makes ByteStreamUploader independent of REAPI, similar to how progressive reads made AbstractRemoteActionCache independent of gRPC.

File: src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java (hunk 2 above)
Comment (werkt) [reply src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java]: I'd rather have newResourceName be deterministic and pure-functional here (will also change its name to just uploadResourceName) - the creation of the upload key should be governed by the caller.

**Comment by werkt** [comment]
Any reason this should be delayed further?

**Comment by buchgr** [comment]
Will import this tomorrow! Sorry about the delay!

**Comment by werkt** [comment]
Merged, hooray! :tada:

**Pull request closed without merging**